sdk/pinocchio/src/cpi.rs

```diff
@@ -13,137 +13,269 @@ use crate::{
-        }
-
-        accounts[index].write(Account::from(account_infos[index]));
-    }
+    _invoke_signed_with_bounds::<ACCOUNTS>(instruction, account_infos, signers_seeds)
+}
 
-    unsafe {
-        invoke_signed_unchecked(
-            instruction,
-            core::slice::from_raw_parts(accounts.as_ptr() as _, ACCOUNTS),
-            signers_seeds,
-        );
+/// Invoke a cross-program instruction with signatures from a slice of
+/// `AccountInfo`s.
+///
+/// This function performs validation of the `account_infos` slice to ensure that:
+///   1. The accounts match the expected accounts in the instruction, i.e., their
+///      `Pubkey` matches the `pubkey` in the `AccountMeta`.
+///   2. The borrow state of the accounts is compatible with the mutability of the
+///      accounts in the instruction.
+///
+/// This validation is done to ensure that the borrow checker rules are followed,
+/// consuming CUs in the process. The `invoke_signed_unchecked` is an alternative
+/// to this function that has lower CU consumption since it does not perform
+/// any validation. This should only be used when the caller is sure that the borrow
+/// checker rules are followed.
+///
+/// The `MAX_ACCOUNTS` constant defines the maximum number of accounts expected
+/// to be passed to the cross-program invocation. This provides an upper bound to
+/// the number of accounts that need to be statically allocated for cases where the
+/// number of instruction accounts is not known at compile time. The final number of
+/// accounts passed to the cross-program invocation will be the number of accounts
+/// required by the `instruction`, even if `MAX_ACCOUNTS` is greater than that.
+///
+/// Note that this function is inlined to avoid the overhead of a function call,
+/// but uses stack memory allocation. When a large number of accounts is needed,
+/// it is recommended to use the [`slice_invoke_signed`] function instead to reduce
+/// stack memory utilization.
+///
+/// # Important
+///
+/// The accounts on the `account_infos` slice must be in the same order as the
+/// `accounts` field of the `instruction`. When the instruction has duplicated
+/// accounts, it is necessary to pass a duplicated reference to the same account
+/// to maintain the 1:1 relationship between `account_infos` and `accounts`.
+#[inline(always)]
+pub fn invoke_signed_with_bounds<const MAX_ACCOUNTS: usize>(
+    instruction: &Instruction,
+    account_infos: &[&AccountInfo],
+    signers_seeds: &[Signer],
+) -> ProgramResult {
+    // Check that the potential number of accounts provided is not less than
+    // the number of accounts expected by the instruction;
+    //
+    // The check for the slice of `AccountInfo`s not being less than the
+    // number of accounts expected by the instruction is done in
+    // `invoke_signed_with_bounds`.
+    if MAX_ACCOUNTS < instruction.accounts.len() {
+        return Err(ProgramError::NotEnoughAccountKeys);
```

Comment:
Sorry for being picky, but I find this comment confusing.
The check is not looking at the number of provided accounts.
It only checks the number of accounts an instruction is expecting.

I think the comment should say something like "make sure that the stack allocated account storage is sufficient for the instruction being used".